raft.go

```diff
@@ -588,54 +588,60 @@ func (r *raft) send(m pb.Message) {
 	}
 }
 
-// sendAppend sends an append RPC with new entries (if any) and the
-// current commit index to the given peer.
-func (r *raft) sendAppend(to uint64) {
-	r.maybeSendAppend(to, true)
-}
-
-// maybeSendAppend sends an append RPC with new entries to the given peer,
-// if necessary. Returns true if a message was sent. The sendIfEmpty
-// argument controls whether messages with no entries will be sent
-// ("empty" messages are useful to convey updated Commit indexes, but
-// are undesirable when we're sending multiple messages in a batch).
+// sendAppend sends an append RPC with new entries to the given peer, if
+// necessary. Returns true if a message was sent.
 //
-// TODO(pav-kv): make invocation of maybeSendAppend stateless. The Progress
-// struct contains all the state necessary for deciding whether to send a
-// message.
-func (r *raft) maybeSendAppend(to uint64, sendIfEmpty bool) bool {
-	pr := r.trk.Progress[to]
-	if pr.IsPaused() {
+// This may send an empty append message (with no entries) if replication to
+// this follower is throttled, or there are no new entries but the commit index
+// for the follower can be bumped.
+func (r *raft) sendAppend(to uint64, pr *tracker.Progress) bool {
+	if pr.State == tracker.StateProbe {
+		return !pr.MsgAppFlowPaused && r.maybeSendAppend(to, pr)
+	} else if pr.State != tracker.StateReplicate {
 		return false
-	}
+	} // only StateReplicate below
+
+	// If there are any pending entries and the inflight tracking is not
+	// saturated, send a regular append message (or snapshot).
+	if pr.Next <= r.raftLog.lastIndex() && !pr.Inflights.Full() {
+		return r.maybeSendAppend(to, pr)
+	}
+	// NB: the commit index is periodically sent in the heartbeat messages, so
+	// technically we don't need the CanBumpCommit clause here to guarantee commit
+	// index convergence on the follower. However, sending it via MsgApp here
+	// allows faster (no heartbeat interval delay) convergence in some cases.
+	if pr.CanBumpCommit(r.raftLog.committed) {
+		return r.maybeSendEmptyAppend(to, pr)
+	}
+	// In a throttled StateReplicate, send an empty append message if we haven't
+	// done so recently.
+	//
+	// We must send periodic appends so that eventually the follower either
+	// accepts or rejects it. If we don't do so, replication can stall if all the
+	// in-flight appends are lost/dropped.
+	return !pr.MsgAppFlowPaused && pr.Match < r.raftLog.lastIndex() &&
```

Comment:
nit: this would read better if you split out the two different conditions for sending an empty message from the logic to do so. Something like:
```go
// NB: the commit index is periodically sent in the heartbeat messages, so
// technically we don't need the CanBumpCommit clause here to guarantee commit
// index convergence on the follower. However, sending it via MsgApp here
// allows faster (no heartbeat interval delay) convergence in some cases.
sendEmpty := pr.CanBumpCommit(r.raftLog.committed)

// In a throttled StateReplicate, send an empty append message if we haven't
// done so recently.
//
// We must send periodic appends so that eventually the follower either
// accepts or rejects it. If we don't do so, replication can stall if all the
// in-flight appends are lost/dropped. 
sendEmpty = sendEmpty || !pr.MsgAppFlowPaused && pr.Match < r.raftLog.lastIndex()
	
if sendEmpty {
	return r.maybeSendEmptyAppend(to, pr)
}
return false
```
Up to you though.

Reply:
The PR has evolved a bit, the is no longer relevant.